Fix unexpected exception when enumerating a completely empty drive root

carlossanlop opened the pull request:
Fixes: https://github.com/dotnet/runtime/issues/33475

This is a regression from .NET Framework, where we were properly handling the edge case where we called `GetDirectories()` on a drive root that didn't even have the System Volume Information folder:
https://referencesource.microsoft.com/#mscorlib/system/io/filesystemenumerable.cs,280

I am not sure we can have a unit test for this case: we would require to have access to an empty drive, acquire ownership of the "System Volume Information" folder, delete it, then ensure we can call `GetDirectories()` without throwing. The SVI directory could get unexpectedly regenerated by Windows between its deletion and the test of `GetDirectories()`.

I tested this locally and the exception is gone. I made sure the SVI folder was not generated by Windows after running my test.

carlossanlop committed:
Fix unexpected exception when enumerating a completely empty drive root.

JeremyKuhne commented on src/libraries/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEnumerator.Win32.cs:
```diff
@@ -45,8 +45,12 @@ private unsafe bool GetData()
                 default:
                     int error = (int)Interop.NtDll.RtlNtStatusToDosError(status);
 
+                    // ERROR_FILE_NOT_FOUND or ERROR_NO_MORE_FILES can be reached when enumerating directories in a completely empty drive root (and there isn't even a System Volume Information folder).
```
We already track `NO_MORE_FILES` above- it is not related to nothing being there, it happens for every directory when it gets to the end of the enumeration (outside of the special case for the no files).

I think it would be better to add the NT status code (`STATUS_NO_SUCH_FILE`) in a case above `default:` (which still keeps the common cases first).  That would be a bit more targeted at this known scenario and keep this condition simpler. The comment would be a bit clearer if it said something like `// STATUS_NO_SUCH_FILE can occur when there are NO files in a volume root (usually there are hidden system files)`.

I did see a few other error codes that can translate to `ERROR_FILE_NOT_FOUND` (by blasting through error codes), but I don't think any of them should be expected here.

carlossanlop committed:
Address PR suggestions:
- Add NTStatus for file not found.
- Use it in the switch case. Since it's a rare case, add it right above 'default'.

carlossanlop commented on src/libraries/Common/src/Interop/Windows/NtDll/Interop.NtStatus.cs:
```diff
@@ -1,4 +1,4 @@
-// Licensed to the .NET Foundation under one or more agreements.
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
Encoding in this file was Western European and VS was complaining. All other files are UTF8, so I switched it.

carlossanlop replied:
There's a weird character hidden somewhere in this file that did not fail compilation locally but is failing in the CI. I'll update the file.

carlossanlop commented on src/libraries/Common/src/Interop/Windows/NtDll/Interop.NtStatus.cs:
```diff
@@ -7,15 +7,16 @@ internal static partial class Interop
-        internal const uint STATUS_NO_MEMORY = 0xC0000017;
-        internal const uint STATUS_OBJECT_NAME_NOT_FOUND = 0xC0000034;
-        internal const uint STATUS_NONE_MAPPED = 0xC0000073;
+        internal const uint STATUS_SUCCESS                = 0x00000000;
```
Sorted by value.

carlossanlop committed:
Remove hidden character causing build failure. It showed up after the file's encoding was changed to UTF8.
